### PR TITLE
[WIP][Kernel] Target parallel optimization

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Currently DGL builtin kernel use edge parallel which causes non-determinism in calculating forward results and backward gradients. A deterministic methodology is required. This PR provides a target parallel strategy: when calculation target point to dst nodes, it parallels execution across different dst nodes; when   calculation target point to edges,  it parallels execution across different edges. This totally eliminate the non-determinism.

#1114 
#1056 Fix this bug here

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by yzh119** [comment]
Windows build bug confirmed: `Fatal error C1002 : compiler is out of heap space in pass 2` ([documentation](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1002?view=vs-2019))

Interestly, caffe/pytorch also had met this problem before, their solution is to split the file:https://caffe2.ai/doxygen-c/html/generic_2_t_h_tensor_math_8cpp_source.html .

Also found in tensorflow windows build: https://github.com/tensorflow/tensorflow/issues/23342, they recommend using bazel instead of cmake.

Also a MXNet windows build problem: https://github.com/apache/incubator-mxnet/issues/13958.

It's said that enlarging precompiled header memory could solve the problem (https://docs.microsoft.com/en-us/cpp/build/reference/zm-specify-precompiled-header-memory-allocation-limit?view=vs-2019), however, this does not work for me. 

Another way is to disable some optimizations by setting `/LTCG:PGOptimize`, however, cmake did not recognize this option (https://stackoverflow.com/questions/47654646/how-to-disable-link-time-code-generation-in-cmake).

**Comment by yzh119** [comment]
Limit the maximum broadcasting dimensionality to 2(https://github.com/dmlc/dgl/blob/master/src/kernel/binary_reduce_common.h#L501-L519) would fix the problem. 4 or 8 doesn't work here.

@jermainewang, @classicsong what's your idea?

**Comment by BarclayII** [comment]
Splitting the CPU reducer source files into 32bit and 64bit reduces compiler memory usage by half and solves the issue.

I guess it's better to have a generator script that generates the `binary_bcast_reduce_${op}_${bits}.cc` files when building, instead of putting all these files in the repo?

**Comment by yzh119** [comment]
This PR is being closed because it's outdated and all functions have been implemented in new PR.
